# Курсовой проект, MDS

## НИУ ВШЭ, 2025-26 учебный год

## Курсовой проект. Exploratory Data Analysis

## Общая информация

__Дата выдачи:__ 11.03.2026

__Защита проекта:__ ~ 20.06.2025

## Формат выполнения

Курсовой проект представляет собой творческое инициативное групповое задание, которое выполняется студентами в командах от 4 до 6 человек. В то же время, следует помнить, что чем больше в команде людей, тем выше предъявляются требования в целом к команде, а также к выполняему ею проекту. Команды разрешается формировать из студентов разных групп.

При оценивании результатов выполнения проекта командой, в обязательном порядке проверяется и учитывается в оценке вклад ***каждого участника*** в общую работу команды. Вследствие этого, крайне важно ответственно подойти к выбору сокомандников, распределению обязанностей внутри команды, а также пресечению наличия "мертвых душ" в команде.

При необходимости, команда большинством голосов имеет право исключить человека из команды *до защиты проекта*, сообщив об этом преподавателю, в случае если такой человек никак не проявлялся в ходе выполнения проекта и являлся как раз таки той самой "мертвой душой". Помните про эту возможность, поскольку наличие "мертвых душ" может негативно повлиять на качество и оценку для всего проекта и, как следствие, в целом для всей команды!

Запись составов команд осуществляется по ссылке в таблице:

https://docs.google.com/spreadsheets/d/1AfaFIfCVB21uwLcatl0LpXGxKqDKCJ3jo7F-7Ji4usg/edit?usp=sharing


## Оценивание и штрафы



Групповой проект оценивается исходя из устной защиты (презентации) командой выполненного ими проекта перед комиссией. Комиссия составляется из преподавателей курса, ассистентов курса, а также приглашенных экспертов из индустрии.

В рамках каждой защиты комиссия выставляет два типа оценок.

Первая оценка — это оценка для всего проекта в целом. Она ставится комиссией, исходя из глубины проработки командой предметной области, сложности применяемых в проекте инструментов, а также прочих аспектов, связанных с содержанием выполняемого в проекте ТЗ. При выставлении этой оценки комиссия ориентируется на критерии, более подробно представленные ниже в разделе "Критерии оценки проекта".

Вторая оценка — это индивидуальная оценка каждого участника команды. Именно эта оценка выставляется впоследствии в ведомость. Индивидуальная оценка участника команды определяется комиссией, исходя из того, насколько участник ориентируется в выполненном командой проекте, как он отвечает на адресованные персонально ему вопросы комиссии, насколько в принципе прослеживается его роль в проекте, а также исходя из других аспектов, связанных с участием конкретного человека в деятельности команды. Индивидуальная оценка не может превосходить оценку команды за проект в целом (первую оценку), однако, в случае когда участник спокойно и корректно отвечает на все адресованные ему вопросы, свободно ориентируется в материале и содержании выполненного проекта, а также когда его роль в работе команды не вызывает сомнений, — в таком случае, индивидуальная оценка участника становится равной оценке для всего проекта (т.е. в идеальном случае — вторая оценка равна первой оценке).

Во избежание субъективности и/или предвзятости оценки тех или иных членов комиссии в отношении тех или иных команд, а также каких-либо иных проявлений неравных условий — распределение команд между комиссиями осуществляется абсолютно рандомно с использованием средств библиотеки Numpy :)) — в качестве рэндом сида выступает дата, в которую производится распределение, например, `np.random.seed(11032026)`.

Для сохранения справедливости и беспристрастности случайного распределения, любые переводы команд между комиссиями (смена комиссии) не допускаются. Присутствие студента на защите проекта является обязательным; вклад студента при его отсутствии не оценивается и автоматически считается равным нулю. Кроме того, во время защиты проекта обязательно, чтобы камера оставалась включённой на протяжении всего времени пребывания в конференции.

## Описание задания

<img src="https://pp.userapi.com/c854528/v854528797/c62ce/FiPEibYUxJc.jpg" width="40%">

Netflix за последние 10-15 лет обзавелись большим количеством зрителей. С увеличением числа зрителей увеличилось и разнообразие шоу. Соответственно, перед аналитиками из киноиндустрии встала задача исследования данных с рейтингами различных сериалов.  

В данном задании вам предстоит провести визуальный анализ датасета **1000 Netflix Shows** (по состоянию на 11.06.2017) и сделать выводы.  



### Описание признаков


* `title` - название шоу.
* `rating` - рейтинг шоу. Например: G, PG, TV-14, TV-MA.
* `ratingLevel` - описание рейтинговой группы и особенностей шоу.
* `ratingDescription` - рейтинг шоу, закодированный числом.
* `release year` - год выпуска шоу.
* `user rating score` - оценка пользователей.
* `user rating size` - общий рейтинг пользователей.

### Требования к проекту

В качестве результата выполнения курсового проекта ваша команда должна получить презентацию и защитить ее перед комиссией.

Оформление презентации остаётся полностью на ваше усмотрение, но помните, что результат должен быть релевантен для демонстрации бизнес-заказчику — комиссию, принимающую вашу работу, правильнее всего воспринимать именно в таком качестве. Например, вставлять в презентацию строчки кода или злоупотреблять скринами блокнота не рекомендуется.

С точки зрения концепции выполнения проекта — вам необходимо принять на себя роль аналитиков: провести работу над признаками, исследовать информацию, содержащуюся в датасете,  выявить тенденции, тренды, факты из данных, а также, конечно, презентовать всё это в понятном виде.

Фактически, можно воспринимать этот проект в следующем ключе: к вам пришел некий бизнес-заказчик — например, это может быть непосредственно представитель Netflix, которые хотят улучшить какие-то процессы; или какая-то компания, которая хочет снять какой-то новый проект и/или продать какой-то свой уже готовый продукт Netflix'у; или это может быть некая компания конкурент Netflix'a, которая заинтересована в общем исследовании рынка; или абсолютно любой другой стейхолдер в рамках данной отрасли — и вот этот заказчик просит вас проанализировать данные и извлечь на основе них какие-то полезные и значимые бизнес-инсайты для него. Разумеется, чем глубже, чем осмысленнее и чем нетривиальнее будут эти выводы, тем больше они вам заплатят :)


### Примеры вопросов и тем


Чтобы вы не начинали работу полностью с чистого листа, мы подобрали для вас некоторое количество опорных вопросов, с которых вы можете начать ваш первичный анализ.

**Внимание:** Эти вопросы лишь *приблизительные* и дают только *начальный вектор* для вашего исследования. Хотя качественно проработанные ответы на них и могут принести вам значительное количество первичных инсайтов и полезной информации — не стоит строить весь ваш проект вокруг только ответов на них. По возможности постарайтесь усовершенствовать и развить темы, затрагивающиеся в данных вопросах, либо вообще найти другие слои, которые в рамках них не рассматриваются (на самом деле, таковых довольно много!)


Итак, сами вопросы:

- Сколько рейтинговых групп представлено в данных?
- Как рейтинговые группы отличаются друг от друга?
- Какие оценки пользователи ставят чаще всего?
- Какие оценки пользователи ставят в среднем?
- В какие годы были запущены шоу?
- Можно ли сделать вывод, что 2017 год успешнее для Netflix, чем 2016?
- Выберите любое шоу (например, самое понравившееся) и составьте его описательный портрет. На какие вопросы о нем вам было бы интересно ответить, если бы вы были заказчиком? Ответьте на эти вопросы :)
- Какие существуют признаки, не представленные в данных, но существенно влияющие на имеющиеся признаки?
- Как формируются оценки?
- Каким образом собирались сами данные?



Повторимся еще раз: это лишь *примерные вопросы*, для того чтобы дать вам начальный вектор, в котором вы могли бы двигаться. Вы можете отвечать или не отвечать на них по своему усмотрению, а также действовать таким образом и в том направлении, как полностью посчитаете нужным!



### Полезные замечания и советы

- Следите за тем, чтобы все графики на слайдах можно было прочитать даже без ваших комментариев;
- Не нужно делать просто текстовые слайды;
- Стройте свой анализ, исходя из того, что ваша цель представить полезную информацию для бизнеса;
- Любой хороший Data Scientist должен уметь работать в команде, поэтому распределение ваших усилий в проекте остается за вами;
- Не обязательно ограничиваться представленным датасетом. У вас есть весь интернет, чтобы найти дополнительную информацию и проверить ваши гипотезы. Дерзайте!


### Примерный скелет EDA

Опять-таки, чтобы вам не приходилось начинать написание кода с чистого листа, мы выстроили некий простейший скелет возможных этапов, которые вы могли бы включить в основу вашего первичного анализа.

Как вы уже, наверное, поняли, мы не настаиваем! :)

#### Скачайте данные
Загрузите данные, выполнив код ниже.

In [335]:
import pandas as pd
import numpy as np
from pathlib import Path

# Чтобы любой мог сразу запустить после клона проекта из гитхаба

project_dir = Path.cwd().parent

raw_data_path = project_dir / "data" / "raw" / "NetflixShows.xlsx"
processed_data_path = project_dir / "data" / "processed" / "NetflixShows_clean.csv"

data_raw = pd.read_excel(raw_data_path)

data_raw.head()

,title,rating,ratingLevel,ratingDescription,release year,user rating score,user rating size
0,White Chicks,PG-13,"crude and sexual humor, language and some drug...",80,2004,82.0,80
1,Lucky Number Slevin,R,"strong violence, sexual content and adult lang...",100,2006,NaN,82
2,Grey's Anatomy,TV-14,Parents strongly cautioned. May be unsuitable ...,90,2016,98.0,80
3,Prison Break,TV-14,Parents strongly cautioned. May be unsuitable ...,90,2008,98.0,80
4,How I Met Your Mother,TV-PG,Parental guidance suggested. May not be suitab...,70,2014,94.0,80


In [336]:
print(f"Размер таблицы: {data_raw.shape}")

Размер таблицы: (1000, 7)


In [337]:
print(f"Названия колонок: {data_raw.columns.tolist()}")

Названия колонок: ['title', 'rating', 'ratingLevel', 'ratingDescription', 'release year', 'user rating score', 'user rating size']


In [338]:
# Смотрим на типы колонок и на то,
# сколько непустых значений

data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   title              1000 non-null   object 
 1   rating             1000 non-null   object 
 2   ratingLevel        941 non-null    object 
 3   ratingDescription  1000 non-null   int64  
 4   release year       1000 non-null   int64  
 5   user rating score  605 non-null    float64
 6   user rating size   1000 non-null   int64  
dtypes: float64(1), int64(3), object(3)
memory usage: 54.8+ KB


In [339]:
# Считаем пропуски

data_raw.isna().sum()

title                  0
rating                 0
ratingLevel           59
ratingDescription      0
release year           0
user rating score    395
user rating size       0
dtype: int64

In [340]:
# Теперь в процентах на всякий, мб для отчета понадобится

miss_count = data_raw.isna().sum()
miss_share = miss_count / data_raw.shape[0] * 100

print("Доля пропусков в процентах:")
miss_share

Доля пропусков в процентах:


title                 0.0
rating                0.0
ratingLevel           5.9
ratingDescription     0.0
release year          0.0
user rating score    39.5
user rating size      0.0
dtype: float64

In [341]:
# Смотрим базовую статистику по числовым колонкам,
# мб сразу что-то вылезет

data_raw.describe().T

,count,mean,std,min,25%,50%,75%,max
ratingDescription,1000.0,67.366000,30.785610,10.0,35.0,60.0,90.0,124.0
release year,1000.0,2010.335000,8.877927,1940.0,2007.0,2015.0,2016.0,2017.0
user rating score,605.0,84.094215,12.344371,55.0,75.0,88.0,95.0,99.0
user rating size,1000.0,80.783000,0.973066,80.0,80.0,80.0,82.0,82.0


In [342]:
# user rating size странный, посмотрим значения отдельно его

print("Значения user rating size:")
data_raw['user rating size'].value_counts()

Значения user rating size:


user rating size
80    605
82    388
81      7
Name: count, dtype: int64

In [343]:
# и ratingDescription тоже,
# потому что похоже на какой-то числовой код рейтинга

print("Значения ratingDescription:")
data_raw['ratingDescription'].value_counts().sort_index()

Значения ratingDescription:


ratingDescription
10      68
35     190
41      38
42      44
60     170
70      59
80      15
90     234
100     19
110    148
124     15
Name: count, dtype: int64

In [344]:
# смотрим, как user rating size связан с оценкой

print("user rating size у строк, где есть user rating score:")
data_raw[data_raw['user rating score'].notna()]['user rating size'].value_counts()

user rating size у строк, где есть user rating score:


user rating size
80    605
Name: count, dtype: int64

In [345]:
# и тут

print("user rating size у строк, где user rating score пропущен:")
data_raw[data_raw['user rating score'].isna()]['user rating size'].value_counts()

user rating size у строк, где user rating score пропущен:


user rating size
82    388
81      7
Name: count, dtype: int64

In [346]:
# Смотрим, является ли ratingDescription просто кодом для рейтинга

data_raw.groupby('rating')['ratingDescription'].value_counts()

rating    ratingDescription
G         35                   138
NR        124                   14
PG        60                   170
PG-13     80                    15
R         100                   19
TV-14     90                   234
TV-G      35                    52
TV-MA     110                  148
TV-PG     70                    59
TV-Y      10                    68
TV-Y7     41                    38
TV-Y7-FV  42                    44
UR        124                    1
Name: count, dtype: int64

После проверки видно, что ratingDescription почти работает как числовой код для rating, поэтому для анализа удобнее оставить обычный rating. user rating size тоже выглядит странно, то есть если user rating score заполнен, то user rating size всегда = 80, а если оценка пропущена, то там стоят 81 или 82, поэтому этот признак больше похож на служебный, решил снести эти колонки

In [347]:
# Делаю копию данных, чтобы с ней работать дальше
# и сношу ненужные служебные колонки

data_work = data_raw.copy()

del data_work['ratingDescription']
del data_work['user rating size']

data_work.head()

,title,rating,ratingLevel,release year,user rating score
0,White Chicks,PG-13,"crude and sexual humor, language and some drug...",2004,82.0
1,Lucky Number Slevin,R,"strong violence, sexual content and adult lang...",2006,NaN
2,Grey's Anatomy,TV-14,Parents strongly cautioned. May be unsuitable ...,2016,98.0
3,Prison Break,TV-14,Parents strongly cautioned. May be unsuitable ...,2008,98.0
4,How I Met Your Mother,TV-PG,Parental guidance suggested. May not be suitab...,2014,94.0


#### Удалите из данных дубликаты.
- Почему они возникли?
- Много ли их? В каких группах их больше всего?

In [348]:
# Смотрим, сколько было строк до удаления

print(f"Строки до удаления дублей: {data_work.shape[0]}")

Строки до удаления дублей: 1000


In [349]:
# Теперь именно полные дубли строк
# значит, если их снесем, то 500 и останется

print(f"Полные дубли строк: {data_work.duplicated().sum()}")

Полные дубли строк: 500


In [350]:
# Смотрим примеры дублей

full_duplicates = data_work[data_work.duplicated() == True]
full_duplicates.head(10)

,title,rating,ratingLevel,release year,user rating score
50,Lucky Number Slevin,R,"strong violence, sexual content and adult lang...",2006,NaN
52,Grey's Anatomy,TV-14,Parents strongly cautioned. May be unsuitable ...,2016,98.0
53,Prison Break,TV-14,Parents strongly cautioned. May be unsuitable ...,2008,98.0
56,How I Met Your Mother,TV-PG,Parental guidance suggested. May not be suitab...,2014,94.0
57,Supernatural,TV-14,Parents strongly cautioned. May be unsuitable ...,2016,95.0
58,Breaking Bad,TV-MA,For mature audiences. May not be suitable for...,2013,97.0
61,The Vampire Diaries,TV-14,Parents strongly cautioned. May be unsuitable ...,2017,91.0
67,The Walking Dead,TV-MA,For mature audiences. May not be suitable for...,2015,98.0
68,Pretty Little Liars,TV-14,Parents strongly cautioned. May be unsuitable ...,2016,96.0
74,Black Mirror,TV-MA,For mature audiences. May not be suitable for...,2016,80.0


In [351]:
# Смотрим, какие названия чаще всего появляются

data_work['title'].value_counts().head(15)

title
13 Reasons Why                        8
Girlboss                              7
Prison Break                          7
New Girl                              6
Pretty Little Liars                   6
The Vampire Diaries                   6
Anastasia                             6
The Secret Life of Pets               6
Dinotrux                              6
Grace and Frankie                     6
The Iron Giant                        6
Shameless (U.S.)                      6
Grey's Anatomy                        6
An American Tail: Fievel Goes West    5
Balto                                 5
Name: count, dtype: int64

In [352]:
# Берем самое частое, плюс видим,
# что 13 Reasons Why прост размножился
# одинаковыми строками

data_work[data_work['title'] == '13 Reasons Why']

,title,rating,ratingLevel,release year,user rating score
41,13 Reasons Why,TV-MA,For mature audiences. May not be suitable for...,2017,99.0
141,13 Reasons Why,TV-MA,For mature audiences. May not be suitable for...,2017,99.0
189,13 Reasons Why,TV-MA,For mature audiences. May not be suitable for...,2017,99.0
241,13 Reasons Why,TV-MA,For mature audiences. May not be suitable for...,2017,99.0
295,13 Reasons Why,TV-MA,For mature audiences. May not be suitable for...,2017,99.0
347,13 Reasons Why,TV-MA,For mature audiences. May not be suitable for...,2017,99.0
396,13 Reasons Why,TV-MA,For mature audiences. May not be suitable for...,2017,99.0
497,13 Reasons Why,TV-MA,For mature audiences. May not be suitable for...,2017,99.0


In [353]:
# Говорили, что датасет хитрый, поэтому
# держим в голове, что одинаковый tile != одинаковая строка

titles = data_work['title'].value_counts()
repeat_titles = titles[titles > 1].index

bad_titles = []

for title in repeat_titles:
    temp = data_work[data_work['title'] == title]
    if temp.drop_duplicates().shape[0] > 1:
        bad_titles.append(title)

bad_titles

['Star Wars: The Clone Wars', 'Goosebumps', 'Bordertown', 'Skins']

In [354]:
# Ручками проверяем, что за названия вылезли

data_work[data_work['title'].isin(bad_titles)].sort_values('title')

,title,rating,ratingLevel,release year,user rating score
167,Bordertown,TV-14,Parents strongly cautioned. May be unsuitable ...,2016,86.0
225,Bordertown,TV-14,Parents strongly cautioned. May be unsuitable ...,2016,86.0
449,Bordertown,TV-MA,For mature audiences. May not be suitable for...,2016,NaN
568,Goosebumps,TV-Y7,Suitable for children ages 7 and older,1998,88.0
632,Goosebumps,PG,"scary and intense creature action and images, ...",2015,90.0
671,Goosebumps,TV-Y7,Suitable for children ages 7 and older,1998,88.0
151,Skins,TV-MA,For mature audiences. May not be suitable for...,2013,NaN
181,Skins,TV-MA,NaN,2017,NaN
504,Star Wars: The Clone Wars,PG,"sci-fi action violence throughout, brief langu...",2008,57.0
512,Star Wars: The Clone Wars,TV-PG,Parental guidance suggested. May not be suitab...,2014,93.0


In [355]:
# Сносим только полные дубли строк, но
# по title не удаляю, потому что записи могут быть разными
# даже с одинаковыми названиями

data_clean = data_work.drop_duplicates().reset_index(drop=True)
print(f"Размер до удаления дублей: {data_work.shape}")

Размер до удаления дублей: (1000, 5)


In [356]:
print(f"Размер после удаления дублей: {data_clean.shape}")

Размер после удаления дублей: (500, 5)


In [357]:
print(f"Полных дублей осталось: {data_clean.duplicated().sum()}")

Полных дублей осталось: 0


После удаления полностью одинаковых строк осталось 500. Они могли появиться из-за повторной выгрузки или объединения одинаковых записей при сборе датасета, больше всего повторов было у тайтлов, например, 13 Reasons Why был 8 раз, Girlboss и Prison Break по 7 раз. При этом удалять строки только по тайтлу нельзя, тк в данных есть одинаковые названия с разными годами, рейтингами или оценками, например Star Wars: The Clone Wars, Goosebumps, Bordertown и Skins, поэтому я удалял только полные дубли строк.

In [358]:
# Снова считаем пропуски заново

miss_count_clean = data_clean.isna().sum()
miss_share_clean = miss_count_clean / data_clean.shape[0] * 100

print("Кол-во пропусков после удаления дублей:")
miss_count_clean

Кол-во пропусков после удаления дублей:


title                  0
rating                 0
ratingLevel           33
release year           0
user rating score    244
dtype: int64

In [359]:
print("Доля пропусков после удаления дублей в процентах:")
miss_share_clean

Доля пропусков после удаления дублей в процентах:


title                 0.0
rating                0.0
ratingLevel           6.6
release year          0.0
user rating score    48.8
dtype: float64

In [360]:
# Смотрим, где нет ratingLevel, но пока не удаляю,
# потому что строки могут быть норм

data_clean[data_clean['ratingLevel'].isna()].head(10)

,title,rating,ratingLevel,release year,user rating score
25,Marvel's Iron Fist,TV-MA,NaN,2017,98.0
31,Opera̵̤es Especiais,TV-MA,NaN,2015,NaN
39,Girlboss,TV-MA,NaN,2017,94.0
42,Sand Castle,TV-MA,NaN,2017,89.0
44,Sandy Wexler,TV-14,NaN,2017,92.0
45,O Amor no Diṿ,TV-MA,NaN,2016,NaN
46,Stronger Than the World,NR,NaN,2016,NaN
48,Happily Married,TV-14,NaN,2015,NaN
49,Pandora,TV-MA,NaN,2016,NaN
72,Ingobernable,TV-MA,NaN,2017,96.0


In [361]:
# Смотрим, где чаще всего нет ratingLevel

data_clean[data_clean['ratingLevel'].isna()]['rating'].value_counts()

rating
TV-MA    22
TV-14     5
NR        3
TV-PG     2
G         1
Name: count, dtype: int64

In [362]:
# Помним про то, что датасет хитрый, проверяем, можно ли восстановить
# ratingLevel через rating, и выясняем, что у некоторых rating куча разных описаний

data_clean.groupby('rating')['ratingLevel'].value_counts()

rating    ratingLevel                                                                                                                  
G         General Audiences. Suitable for all ages.                                                                                        52
NR        This movie has not been rated.                                                                                                    7
PG        Parental guidance suggested. May not be suitable for children.                                                                   12
          mild thematic elements                                                                                                            3
          action and rude humor                                                                                                             1
                                                                                                                                           ..
TV-PG     Pa

In [363]:
# считаем, сколько разных ratingLevel есть внутри кждого rating

for rating in data_clean['rating'].unique():
    temp = data_clean[data_clean['rating'] == rating]['ratingLevel'].dropna()
    
    print(rating, "-", len(temp.unique()))

PG-13 - 12
R - 14
TV-14 - 2
TV-PG - 1
TV-MA - 1
TV-Y - 1
NR - 1
TV-Y7-FV - 1
UR - 1
PG - 63
TV-G - 1
G - 1
TV-Y7 - 1


In [364]:
# теперь смотрим, где чаще всего нет rating score

print("Где user rating score пропущен:")
data_clean[data_clean['user rating score'].isna() == True]['rating'].value_counts()

Где user rating score пропущен:


rating
TV-MA       42
G           34
TV-14       29
PG          28
TV-Y        28
TV-G        18
TV-Y7       17
TV-Y7-FV    17
TV-PG       12
NR           8
R            7
PG-13        3
UR           1
Name: count, dtype: int64

In [365]:
print("Где user rating score есть:")
data_clean[data_clean['user rating score'].isna() == False]['rating'].value_counts()

Где user rating score есть:


rating
TV-14       77
PG          48
TV-MA       40
TV-PG       21
G           19
TV-G        11
PG-13        9
TV-Y         8
TV-Y7-FV     8
R            7
TV-Y7        6
NR           2
Name: count, dtype: int64

После удаления дублей пропуски остались только в ratingLevel и user rating score.

ratingLevel не заполняю, потому что у некоторых рейтинговых групп есть много разных текстовых описаний. Например, внутри PG, R, PG-13 встречаются разные варианты описаний, а искажать данные мы не хотим

user rating score тоже не заполняю и не удаляю строки с пропусками, оценки юзеров нет у 244 строк из 500, то есть почти у половины очищенного датасета. Если удалить эти строки, мы потеряем слишком много данных, а если заменить, то создадим искусственные оценки, поэтому пропуски оставлены как NaN, а при дальнейшем анализе оценок нужно учитывать, что оценки есть только у 256 строк.

### Финальная проверка очищенной таблицы

In [366]:
# Финально смотрим итоговую структуру

data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   title              500 non-null    object 
 1   rating             500 non-null    object 
 2   ratingLevel        467 non-null    object 
 3   release year       500 non-null    int64  
 4   user rating score  256 non-null    float64
dtypes: float64(1), int64(1), object(3)
memory usage: 19.7+ KB


In [367]:
# Теперь стату уже по очищенной таблицу

data_clean.describe().T

,count,mean,std,min,25%,50%,75%,max
release year,500.0,2010.976000,7.829305,1940.0,2009.0,2014.0,2016.0,2017.0
user rating score,256.0,81.398438,12.730904,55.0,71.0,83.5,93.0,99.0


In [368]:
# Ну и дальше смотрим, сколько чего осталось

print(f"Уникальных названий: {data_clean['title'].nunique()}")

Уникальных названий: 496


In [369]:
print(f"Уникальных рейтинговых групп: {data_clean['rating'].nunique()}")

Уникальных рейтинговых групп: 13


In [370]:
print(f"Уникальных годов выпуска: {data_clean['release year'].nunique()}")

Уникальных годов выпуска: 35


In [371]:
# Смотрим распределение по рейтинговым группам

data_clean['rating'].value_counts()

rating
TV-14       106
TV-MA        82
PG           76
G            53
TV-Y         36
TV-PG        33
TV-G         29
TV-Y7-FV     25
TV-Y7        23
R            14
PG-13        12
NR           10
UR            1
Name: count, dtype: int64

In [372]:
# И годы выпуска по порядку

data_clean['release year'].value_counts().sort_index()

release year
1940      1
1976      1
1978      1
1982      2
1986      1
1987      1
1989      2
1990      2
1991      1
1992      1
1993      5
1994      4
1995      2
1996      1
1997      9
1998      9
1999      5
2000      7
2001      6
2002      5
2003     11
2004      7
2005     10
2006     10
2007      8
2008     11
2009     10
2010     13
2011     15
2012     27
2013     33
2014     35
2015     61
2016    146
2017     37
Name: count, dtype: int64

In [373]:
# Самые часты года выпуска

data_clean['release year'].value_counts().head(10)

release year
2016    146
2015     61
2017     37
2014     35
2013     33
2012     27
2011     15
2010     13
2008     11
2003     11
Name: count, dtype: int64

In [374]:
# Помню, что в выводе видел кракозябры, наверное,
# из-за кривой кодировке, поэтому проверим и их тоже

bad_symbols = ['Ì', 'Ò', 'À', '¤', 'µ', '©']
bad_names = []

for title in data_clean['title'].unique():
    title_text = str(title)

    for symbol in bad_symbols:
        if symbol in title_text:
            bad_names.append(title_text)
            break

bad_names

['ÌÒ Pai, ÌÒ - Look at This',
 'Opera̵̤es Especiais',
 'O Amor no Diṿ',
 'MerlÌ?',
 'Pok̩mon: Indigo League',
 'Pok̩mon: XY']

In [375]:
# И еще на всякий смотрим, что все title действительно строки

for title in data_clean['title'].unique():
    if type(title) != str:
        print(title, type(title))

0.03 <class 'float'>
90210 <class 'int'>


In [376]:
# А у меня тоже есть птички... Собираю странные title в список
# чтобы посмотреть эти строки целиком

bad_title_values = []

for title in data_clean['title'].unique():
    title_text = str(title)
    for symbol in bad_symbols:
        if symbol in title_text:
            bad_title_values.append(title)
            break
    
    if type(title) != str:
        bad_title_values.append(title)

bad_title_values

['ÌÒ Pai, ÌÒ - Look at This',
 'Opera̵̤es Especiais',
 0.03,
 'O Amor no Diṿ',
 90210,
 'MerlÌ?',
 'Pok̩mon: Indigo League',
 'Pok̩mon: XY']

In [377]:
# Теперь смотрим те самые строки

data_clean[data_clean['title'].isin(bad_title_values)]

,title,rating,ratingLevel,release year,user rating score
21,"ÌÒ Pai, ÌÒ - Look at This",TV-MA,For mature audiences. May not be suitable for...,2007,NaN
31,Opera̵̤es Especiais,TV-MA,NaN,2015,NaN
32,0.03,TV-MA,For mature audiences. May not be suitable for...,2016,NaN
45,O Amor no Diṿ,TV-MA,NaN,2016,NaN
51,90210,TV-14,Parents strongly cautioned. May be unsuitable ...,2013,62.0
113,MerlÌ?,TV-PG,Parental guidance suggested. May not be suitab...,2015,NaN
325,Pok̩mon: Indigo League,TV-Y,Suitable for all ages.,2000,74.0
339,Pok̩mon: XY,TV-Y7-FV,Suitable for children ages 7 and older. Conte...,2015,56.0


In [378]:
# Далее привожу title к тексту, при этом 90210 пусть остается названием,
# а 0.03 поменяю на 3%, потому что эксель прочел название как процент

data_clean['title'] = data_clean['title'].astype(str)

data_clean.loc[data_clean['title'] == '0.03', 'title'] = '3%'

data_clean[data_clean['title'].isin(['3%', '90210'])]

,title,rating,ratingLevel,release year,user rating score
32,3%,TV-MA,For mature audiences. May not be suitable for...,2016,NaN
51,90210,TV-14,Parents strongly cautioned. May be unsuitable ...,2013,62.0


In [379]:
# Смотрим, остались ли еще какие-нибудь title,
# которые не являются строками

for title in data_clean['title'].unique():
    if type(title) != str:
        print(title, type(title))

In [380]:
print(f"Размер таблицы: {data_clean.shape}")

Размер таблицы: (500, 5)


In [381]:
print(f"Полных дублей после правки title: {data_clean.duplicated().sum()}")

Полных дублей после правки title: 0


In [382]:
print(f"Уникальных названий после правки title: {data_clean['title'].nunique()}")

Уникальных названий после правки title: 496


In [383]:
print(f"Колонки: {data_clean.columns.tolist()}")

Колонки: ['title', 'rating', 'ratingLevel', 'release year', 'user rating score']


In [384]:
print("Пропуски:")
data_clean.isna().sum()

Пропуски:


title                  0
rating                 0
ratingLevel           33
release year           0
user rating score    244
dtype: int64

In [385]:
data_clean.sample(10)

,title,rating,ratingLevel,release year,user rating score
175,Twilight,PG-13,some violence and one sexual scene,2008,89.0
332,Some Assembly Required,TV-G,Suitable for all ages.,2015,NaN
302,Lego DC Comics: Batman Be-Leaguered,TV-Y7,Suitable for children ages 7 and older,2014,NaN
272,Thomas & Friends: Thomas & the Runaway Kite,TV-Y,Suitable for all ages.,2010,NaN
220,Marvel's Jessica Jones,TV-MA,For mature audiences. May not be suitable for...,2015,83.0
147,Boy Bye,TV-MA,NaN,2016,NaN
408,Paulie,PG,Parental guidance suggested. May not be suitab...,1998,NaN
31,Opera̵̤es Especiais,TV-MA,NaN,2015,NaN
421,An American Tail: The Mystery of the Night Mon...,G,General Audiences. Suitable for all ages.,1999,NaN
480,Air Bud: World Pup,G,General Audiences. Suitable for all ages.,2000,NaN


In [386]:
# Записываем все для следующего этапа

data_clean.to_csv(processed_data_path, index=False)

In [387]:
# Быстрый чек строчек

check = pd.read_csv(processed_data_path)

print(f"Размер сохраненного файла: {check.shape}")

check.head()

Размер сохраненного файла: (500, 5)


,title,rating,ratingLevel,release year,user rating score
0,White Chicks,PG-13,"crude and sexual humor, language and some drug...",2004,82.0
1,Lucky Number Slevin,R,"strong violence, sexual content and adult lang...",2006,NaN
2,Grey's Anatomy,TV-14,Parents strongly cautioned. May be unsuitable ...,2016,98.0
3,Prison Break,TV-14,Parents strongly cautioned. May be unsuitable ...,2008,98.0
4,How I Met Your Mother,TV-PG,Parental guidance suggested. May not be suitab...,2014,94.0


Итоговая таблица для следующего лежит в NetflixShows_clean.csv. В ней осталось 500 строк и 5 колонок: title, rating, ratingLevel, release year и user rating score. Полных дубли почистил, в тайтл все значения приведены к строковому типу, 0.03 исправлено на 3%, потому что эксель, будь он неладен, при этом названия с кривой кодировкой я не исправлял руками, чтобы случайно не переименовать их неправильно

Крч, user rating score заполнен только у 256 строк из 500, а ratingLevel имеет 33 пропуска - эт ключевое.

#### Сколько рейтинговых групп представлено в данных?

In [388]:
...

Ellipsis

#### Какие оценки пользователи ставят чаще?

- Постройте гистограмму оценок.
- Почему именно такие оценки ставят чаще всего?

In [389]:
...

Ellipsis

#### Выведите основную информацию об оценках пользователей
- Чему равны среднее, стандартное отклонение, минимум, максимум, медиана?
- Отличаются ли медиана и среднее?
- Могут ли данные характеристики значительно отличаться? Почему?

In [390]:
...

Ellipsis

#### В какие годы были запущены шоу, представленные в датасете?

In [391]:
...

Ellipsis

#### Постройте график, показывающий распределение количества запущенных шоу в зависимости от года.

- Наблюдается ли рост?
- Есть ли выбросы?
- Что из себя представляют выбросы?
- Чем могут быть они вызваны?

In [392]:
...

Ellipsis

#### Сравните среднюю оценку пользователей в 2016 со средней оценкой в 2017.
- Можно ли сделать вывод, что 2017 год успешнее для Netflix? ("Успешнее" значит, что пользователи в среднем ставили более высокие оценки)
- Как еще можно оценить "Успешность"?



In [393]:
...

Ellipsis

#### Как нагляднее будет показать распределение пользователей по рейтинговым группам?

Ниже представлены два графика, показывающие распределение шоу по рейтинговым группам. Какой тип графика визуально более интерпретируемый? ([Подсказка](https://sun9-40.userapi.com/c854228/v854228652/c754f/j6z5gMjJy2k.jpg))  
Постройте самостоятельно график, который считаете более интерпретируемым. Сделайте вывод.

Может быть, можно использовать и какие-то другие, еще более продвинутые визуализации?


![Charts](https://pp.userapi.com/c852024/v852024281/1a53b1/jSOsBIhxK3U.jpg)

In [394]:
...

Ellipsis

#### Теперь вам нужно выбрать любое из представленных шоу и при помощи изученных инструментов составить описательный портрет этого шоу :)

Аналитики и Data Scientist-ы очень любят все сравнивать, поэтому при составлении описательного портрета рекомендуем вам сравнить выбранное шоу с другими по каким-либо характеристикам.

In [395]:
...

Ellipsis

## Критерии оценки проекта

Оценка за курсовой проект в целом — то есть первая из двух оценок — выставляется по 10 балльной шкале, исходя из 5 критериев. Также комиссия на своё усмотрение может добавить (но не снизить) некоторое количество бонусных баллов за рамками данных критериев.

Каждый из критериев будет перечислен и рассмотрен ниже

### Критерий 1. Информативность техническая – необходимо выявить нетривиальные зависимости и тенденции в данных (до 2 баллов)

Данный критерий оценивает, насколько вы продвинулись технически с точки зрения выявления каких-то паттернов, тенденций и зависимостей в данных. Чем более сложные инструменты вы примените в ходе выполнения проекта и чем более сложные и нетривиальные зависимости получите, тем выше будет оценка у вашей команды по данному критерию!

### Критерий 2. Информативность аналитическая – необходимо интерпретировать данные и делать выводы полезные для бизнеса  (до 2 баллов)

Данный критерий оценивает прикладной смысл проделанных вами действий, интерпретацию технических инструментов, умение подсветить полезность и значимость полученных выводов с точки зрения бизнеса, умение переложить полученные результаты на исходные бизнес-цели и "продать" то, что вы сделали.


### Критерий 3. Визуализация – оценивается техническое исполнение графиков (до 2 баллов)

Данный критерий оценивает сложность и техничность ваших визуализаций. Визуализации должны быть наглядными, красивыми, понятными, легко-интерпертируемыми, нетривиальными с точки зрения технического исполнения и выводов, которые они несут в себе.


*Максимальный балл по данному критерию (**2 балла**) нельзя получить, если в работе используются лишь базовые библиотеки визуализаций: Matplotlib и Seaborn*

### Критерий 4. Предобработка датасета – очистка/замена данных (до 1 балла)

Данный критерий оценивает полноту всех проделанных действий с точки зрения подготовки данных и первичной предобработки датасета.


### Критерий 5. Обогащение данными – добавление новых признаков/датасетов (до 3 баллов)

Данный критерий оценивает полноту использования средств feature engineering на основе имеющихся данных, а также средств обогащения данных путем дополнения имеющегося датасета значимой информацией из связанных сторонних датасетов, различных внешних ресурсов и других источников. Обратите внимание, что связанная информация извне не должна замещать или подменять собой исходную задачу и исходные данные в представленном датасете, а должна лишь дополнять и обогощать его. Корректность и глубина проработки взаимодействия со внешними ресурсами и оценивается в рамках данного критерия!

*Максимальный балл по данному критерию (**3 балла**) можно получить только при условии корректной и значимой работы с как минимум двумя источниками внешних данных (двумя внешними датасетами), а также при условии наличия как минимум двух примеров валидного и осмысленного применения средств feature engineering*

*По данному критерию можно получить **2 балла** только при условии корректной и значимой работы с как минимум одним источником внешних данных (внешним датасетом), а также при условии наличия как минимум одного примера валидного и осмысленного применения средств feature engineering*